django_ses/views.py

```diff
@@ -402,6 +320,243 @@ def handle_event(request, signal_sender=None):
     return HttpResponse()
 
 
-@require_POST
-def handle_bounce(request):
-    return handle_event(request, signal_sender='handle_bounce')
+@method_decorator(csrf_exempt, name='dispatch')
+class SESEventWebhookView(View):
+    """
+    Handle a email sending event via an SNS webhook.
+
+    Parse the event message and send the appropriate signal.
+    <eventType> -> <signal>
+    bounce -> bounce_received
+    complaint -> complaint_received
+    delivery -> delivery_received
+    send -> send_received
+    open -> open_received
+    click -> click_received
+    See: https://docs.aws.amazon.com/ses/latest/DeveloperGuide/monitor-sending-activity.html
+    See: http://docs.aws.amazon.com/sns/latest/gsg/json-formats.html#http-subscription-confirmation-json
+    See: http://docs.amazonwebservices.com/ses/latest/DeveloperGuide/NotificationsViaSNS.html
+
+    In addition to email bounce requests this endpoint also supports the SNS
+    subscription confirmation request. This request is sent to the SNS
+    subscription endpoint when the subscription is registered.
+    See: http://docs.aws.amazon.com/sns/latest/gsg/Subscribe.html
+
+    For the format of the SNS subscription confirmation request see this URL:
+    http://docs.aws.amazon.com/sns/latest/gsg/json-formats.html#http-subscription-confirmation-json
+
+    SNS message signatures are verified by default. This functionality can
+    be disabled by setting AWS_SES_VERIFY_EVENT_SIGNATURES to False.
+    However, this is not recommended.
+    See: http://docs.amazonwebservices.com/sns/latest/gsg/SendMessageToHttp.verify.signature.html
+    """
+
+    def post(self, request, *args, **kwargs):
+        raw_json = request.body
+
+        try:
+            notification = json.loads(raw_json.decode('utf-8'))
+        except ValueError as e:
+            # TODO: What kind of response should be returned here?
+            logger.warning(u'Received bounce with bad JSON: "%s"', e)
+            return HttpResponseBadRequest()
+
+        # Verify the authenticity of the event message.
+        if settings.VERIFY_EVENT_SIGNATURES and not utils.verify_event_message(notification):
+            # Don't send any info back when the notification is not
+            # verified. Simply, don't process it.
+            logger.info(
+                u'Received unverified notification: Type: %s',
+                notification.get('Type'),
+                extra={
+                    'notification': notification,
+                },
+            )
+            return HttpResponse()
```

Comment:
* Could there be a message here that the signature is invalid?
* Could the status code be 400 or will that cause problems with SNS?

Reply:
I think, If the sender is SNS, it should never happen. It returns 400, if the request body is invalid. So it can return 400 as well if the signature is invalid.